components/places/src/import/ios/history.rs

```diff
@@ -0,0 +1,151 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use std::time::Instant;
+
+use crate::api::places_api::PlacesApi;
+use crate::bookmark_sync::engine::update_frecencies;
+use crate::error::Result;
+use crate::import::common::{
+    attached_database, define_history_migration_functions, select_count, HistoryMigrationResult,
+};
+use url::Url;
+
+/// This import is used for iOS users migrating from `browser.db`-based
+/// history storage to the new rust-places store.
+///
+/// The goal of this import is to persist all local browser.db items into places database
+///
+///
+/// ### Basic process
+///
+/// - Attach the iOS database.
+/// - Slurp records into a temp table "iOSHistoryStaging" from iOS database.
+///   - This is mostly done for convenience, to punycode the URLs and some performance benefits over
+///     using a view or reading things into Rust
+/// - Add any entries to moz_places that are needed (in practice, most are
+///   needed, users in practice don't have nearly as many bookmarks as history entries)
+/// - Use iosHistoryStaging and the browser.db to migrate visits to the places visits table.
+/// - Update frecency for new items.
+/// - Cleanup (detach iOS database, etc).
+pub fn import(
+    places_api: &PlacesApi,
+    path: impl AsRef<std::path::Path>,
+) -> Result<HistoryMigrationResult> {
+    let url = crate::util::ensure_url_path(path)?;
+    do_import(places_api, url)
+}
+
+fn do_import(places_api: &PlacesApi, ios_db_file_url: Url) -> Result<HistoryMigrationResult> {
+    let conn_mutex = places_api.get_sync_connection()?;
```

Comment:
If we end up trying to migrate the "live" database this way, we might end up with any other "write" operations failing with SQLITE_BUSY - attempts to create bookmarks, for example. That *might* mean we would be better off using the normal writer connection, so we just block the attempt to create the bookmark until the import is complete. We probably need think more about this as we refine the strategy we end up with.